server/src/test/java/org/elasticsearch/common/settings/ConsistentSettingsIT.java

```diff
@@ -0,0 +1,289 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.settings;
+
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting.AffixSetting;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.test.ESIntegTestCase;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST)
+public class ConsistentSettingsIT extends ESIntegTestCase {
+
+    static final Setting<SecureString> DUMMY_STRING_CONSISTENT_SETTING = SecureSetting
+            .secureString("dummy.consistent.secure.string.setting", null, Setting.Property.Consistent);
+    static final AffixSetting<SecureString> DUMMY_AFIX_STRING_CONSISTENT_SETTING = Setting.affixKeySetting(
+            "dummy.consistent.secure.string.afix.setting.", "suffix",
+            key -> SecureSetting.secureString(key, null, Setting.Property.Consistent));
+    static final Setting<SecureString> EXTRA_DUMMY_STRING_CONSISTENT_SETTING = SecureSetting
+            .secureString("extra.dummy.consistent.secure.string.setting", null, Setting.Property.Consistent);
+    static final AffixSetting<SecureString> DUMMY_AFIX_STRING_INCONSISTENT_SETTING = Setting.affixKeySetting(
+            "dummy.inconsistent.secure.string.afix.setting.", "suffix",
+            key -> SecureSetting.secureString(key, null));
+    private final AtomicReference<Function<Integer, Settings>> nodeSettingsOverride = new AtomicReference<>(null);
+
+    public void testAllConsistentSuccess() throws Exception {
+        internalCluster().getInstances(Environment.class).forEach(environment -> {
+            ClusterService clusterService = internalCluster().getInstance(ClusterService.class);
+            assertTrue(new ConsistentSettingsService(environment.settings(), clusterService, Collections.emptyList()).areAllConsistent());
```

Comment:
No, they're legit integration tests IMO.
They couple the hashes publication with the cluster formation, also starting new nodes. They are end to end tests, where we publish from one node and verify on another.
I am a bit proud of them. The unit tests are in `ConsistentSettingsServiceTests` where we intercept the cluster state and do checks on that.

Reply:
It doesn't seem like you are actually using the nodes really, except to do the creation of ctor args. I would expect an integ test to check the node as it is, but all these assertions construct a service that should be running inside the node. They look like unit tests where you want to tweak the state in eg clusterService through mocks, which is much lighter weight than spinning up even in memory nodes.

Reply:
They do not tweak the cluster state. They look odd because there is no API for this functionality yet, so I have to construct the `ConsistentSettingsService` from the node's `ClusterService` as an API would need to do it. The `ConsistentSettingsService` ctr args are important to be on different nodes (the tests iterates over all of them).

Reply:
Not sure what you mean about different nodes. The ctor args are always going to come from the local node. I don't think there is any reason all of the state that would be provided by different nodes couldn't be mocked. This would make the tests much faster, and easier to debug (no inter node communication).

Reply:
We discussed on another channel and agreed that integration tests should be much thinner.

@rjernst I have pared down `ConsistentSettingsIT`; it contains a test for the success validation and another test for the failed validation, when a new node joins with different settings. The failed validation scenarios (of which there are two) are randomized, because all cases should be covered in `ConsistentSettingsServiceTests`.